evm/src/fuzz/invariant/mod.rs

```diff
@@ -0,0 +1,270 @@
+                    Some(func),
+                    test_contract_abi,
+                    &result,
+                    calldata,
+                    &inner_sequence,
+                )),
+            );
+            found_case = true;
+        }
+    }
+
+    if found_case {
+        eyre::bail!("");
+    }
+    Ok(())
+}
+
+/// The outcome of an invariant fuzz test
+#[derive(Debug)]
+pub struct InvariantFuzzTestResult {
+    pub invariants: BTreeMap<String, Option<InvariantFuzzError>>,
+    /// Every successful fuzz test case
+    pub cases: Vec<FuzzedCases>,
+    /// Number of reverted fuzz calls
+    pub reverts: usize,
+}
+
+#[derive(Debug, Clone)]
+pub struct InvariantFuzzError {
+    /// The proptest error occurred as a result of a test case.
+    pub test_error: TestError<Vec<BasicTxDetails>>,
+    /// The return reason of the offending call.
+    pub return_reason: Reason,
+    /// The revert string of the offending call.
+    pub revert_reason: String,
+    /// Address of the invariant asserter.
+    pub addr: Address,
+    /// Function data for invariant check.
+    pub func: Option<ethers::prelude::Bytes>,
+    /// Inner fuzzing Sequence coming from overriding calls.
+    pub inner_sequence: Vec<Option<BasicTxDetails>>,
+}
+
+impl InvariantFuzzError {
+    fn new(
+        invariant_address: Address,
+        error_func: Option<&Function>,
+        abi: &Abi,
+        result: &bytes::Bytes,
+        calldata: &[BasicTxDetails],
+        inner_sequence: &[Option<BasicTxDetails>],
+    ) -> Self {
+        let mut func = None;
+        let origin: String;
+
+        if let Some(f) = error_func {
+            func = Some(f.short_signature().into());
+            origin = f.name.clone();
+        } else {
+            origin = "Revert".to_string();
+        }
+
+        InvariantFuzzError {
+            test_error: proptest::test_runner::TestError::Fail(
+                format!(
+                    "{}, reason: '{}'",
+                    origin,
+                    match foundry_utils::decode_revert(result.as_ref(), Some(abi)) {
+                        Ok(e) => e,
+                        Err(e) => e.to_string(),
+                    }
+                )
+                .into(),
+                calldata.to_vec(),
+            ),
+            return_reason: "".into(),
+            // return_reason: status,
+            revert_reason: foundry_utils::decode_revert(result.as_ref(), Some(abi))
+                .unwrap_or_default(),
+            addr: invariant_address,
+            func,
+            inner_sequence: inner_sequence.to_vec(),
+        }
+    }
+}
+
+/// Given a TestRunner and a strategy, it generates calls. Used inside the Fuzzer inspector to
+/// override external calls to test for potential reentrancy vulnerabilities..
+#[derive(Debug, Clone)]
+pub struct RandomCallGenerator {
+    /// Runner that will generate the call from the strategy.
+    pub runner: Arc<RwLock<TestRunner>>,
```

Comment:
why do we need shared exclusive access here?

Reply:
In-between `raw_calls` the `InspectorStack` is copied, which clones our `RandomCallGenerator`. If we don't put it in a shared ref, then the cloned `TestRunner` will always be using the starting rng seed in all calls. By sharing it, we make sure that the `rng` evolves in-between calls and does not reset.

The reason behind `Arc` is: https://github.com/foundry-rs/foundry/pull/1572#discussion_r928161235

Reply:
needs docs for reasoning, was confusing at first,
but this still doesn't make perfect sense, does this mean all invariant runners use the same Randomcallgenerator?

wouldn't be solution to create the strategy then when it's needed?